Cpp_hands_on_programming/lib/pipe/Ipc_pipe.cc

```diff
@@ -56,24 +82,40 @@ void Ipc_pipe::receive() {
 	off_t fileSize(0), total(0);
 	std::vector<char> buffer(PIPE_BUF);
 	// open pipe 
-	while ((fdp = open(name.c_str(), O_RDONLY)) == -1) {
-		std::this_thread::sleep_for (std::chrono::seconds(1));
-	}
+	t.startTimer();
+	do {
+		errno = 0;
+		fdp = open(name.c_str(), O_RDONLY | O_NONBLOCK); 
+		t.checkTimer("Cannot connect to pipeSend. Errno: " + std::string(strerror(errno)), "Connecting pipeSend..");
+	} while(errno == ENOENT);
 	// Start read and write loop
-	while ((bytes = read(fdp, buffer.data(), PIPE_BUF)) != 0) {	
-		if (bytes == -1) {
-			throw(std::runtime_error("Ipc_pipe::receive: read"));
-		}
+	std::cout << "Connected!" << std::endl;
+	// Wait for write end to open the pipe
+	std::this_thread::sleep_for (std::chrono::milliseconds(1));
```

Comment:
It is for making sure the write end will not write everything and close the write end before the read end reads the pipe for the first time, it can happen for very small file

Reply:
Ah ok it's because you open the fifo in O_NONBLOCK I think. Maybe you can check the error EWOULDBLOCK after trying to read:
```
EAGAIN or EWOULDBLOCK
              The file descriptor fd refers to a socket and has been
              marked nonblocking (O_NONBLOCK), and the read would block.
              POSIX.1-2001 allows either error to be returned for this
              case, and does not require these constants to have the
              same value, so a portable application should check for
              both possibilities.
```

Reply:
I think 1ms is sufficient because writer will immediately go to the writing loop after read end successfully open the pipe, I haven't tried EWOULDBLOCK, will check. I think this case would have happened even if FIFO  is opened without O_NONBLOCK flag(if the loop exit condition is read != 0), read will return 0 with or without the flag if write end is closed at that point

Reply:
You can be sure about that: the 2 programs can be in different cores so assuming they are in the same RR is a bad idea.

If the fifo is opened without O_NONBLOCK the read() will block until there is data in the pipe or the sender disconnect and in this case an EOF flag will be given to the receiver.

Reply:
Good point, I agree programs could behave differently in such case. 

I don't think the read end will block even O_NONBLOCK is not set. I am talking about the scenario where receiver immediately goes to read the pipe after the sender call mkfifo(), at this point it is possible that the sender has not made it to open() yet, so write end is still closed, then read will return 0 with or without O_NONBLCOK

Reply:
```The kernel maintains exactly one pipe object for each FIFO
       special file that is opened by at least one process.  The FIFO
       must be opened on both ends (reading and writing) before data can
       be passed.  Normally, opening the FIFO blocks until the other end
       is opened also.```
from https://man7.org/linux/man-pages/man7/fifo.7.html

So without non_blocking open() will just block if there is no other process on the other side.

Reply:
Oh sorry you are right,  I forgot read end will block at open() function,

Reply:
> Ah ok it's because you open the fifo in O_NONBLOCK I think. Maybe you can check the error EWOULDBLOCK after trying to read:
> 
> ```
> EAGAIN or EWOULDBLOCK
>               The file descriptor fd refers to a socket and has been
>               marked nonblocking (O_NONBLOCK), and the read would block.
>               POSIX.1-2001 allows either error to be returned for this
>               case, and does not require these constants to have the
>               same value, so a portable application should check for
>               both possibilities.
> ```
> 
> Because 1ms you can't be sure that it will be sufficient for the other thread to start.

It seems to work only with socket not pipe

Reply:
Hmm you're right.

Maybe, you can check in the receiver if the read size is 0 bytes, that means he should try again?

Or maybe the sleep is not the worst idea ^^ (but 1ms is not enough to be sure if programs are running in different cores).